libc/src/__support/FPUtil/generic/div.h

```diff
@@ -104,120 +94,33 @@ div(InType x, InType y) {
 
-  if (q_exponent < -OutFPBits::EXP_BIAS - OutFPBits::FRACTION_LEN) {
-    set_errno_if_required(ERANGE);
-    raise_except_if_required(FE_UNDERFLOW | FE_INEXACT);
-
-    switch (quick_get_round()) {
-    case FE_DOWNWARD:
-      if (result_sign.is_pos())
-        return OutFPBits::zero(result_sign).get_val();
-      return OutFPBits::min_subnormal(result_sign).get_val();
-    case FE_UPWARD:
-      if (result_sign.is_pos())
-        return OutFPBits::min_subnormal(result_sign).get_val();
-      return OutFPBits::zero(result_sign).get_val();
-    default:
-      return OutFPBits::zero(result_sign).get_val();
-    }
-  }
-
-  InStorageType q = 1;
-  InStorageType xd_mant_in = static_cast<InStorageType>(
-      xd.mantissa >> (DYADIC_EXTRA_MANTISSA_LEN - would_q_be_subnormal));
-  InStorageType yd_mant_in =
-      static_cast<InStorageType>(yd.mantissa >> DYADIC_EXTRA_MANTISSA_LEN);
-  InStorageType r = xd_mant_in - yd_mant_in;
-
-  for (size_t i = 0; i < InFPBits::FRACTION_LEN + 1; i++) {
+  for (size_t i = 0; i < NUM_ITERS; ++i) {
     q <<= 1;
-    InStorageType t = r << 1;
-    if (t < yd_mant_in) {
-      r = t;
-    } else {
+    r <<= 1;
+    if (r >= yd_mant_in) {
       q += 1;
-      r = t - yd_mant_in;
+      r -= yd_mant_in;
     }
   }
 
-  bool round;
-  bool sticky;
-  OutStorageType result;
-
-  if (q_exponent > -OutFPBits::EXP_BIAS) {
-    // Result is normal.
-
-    InStorageType round_mask = InStorageType(1) << (Q_EXTRA_FRACTION_LEN - 1);
-    round = (q & round_mask) != 0;
-    InStorageType sticky_mask = round_mask - 1;
-    sticky = (q & sticky_mask) != 0;
-
-    result = OutFPBits::create_value(
-                 result_sign,
-                 static_cast<OutStorageType>(q_exponent + OutFPBits::EXP_BIAS),
-                 static_cast<OutStorageType>(q >> Q_EXTRA_FRACTION_LEN))
-                 .uintval();
-
-  } else {
-    // Result is subnormal.
+  DyadicFloat result(result_sign, result_exp, q);
+  result.mantissa += r != 0;
 
-    // +1 because the leading bit is now part of the fraction.
-    int extra_fraction_len =
-        Q_EXTRA_FRACTION_LEN + 1 - q_exponent - OutFPBits::EXP_BIAS;
+  OutType output = static_cast<OutType>(result);
 
-    InStorageType round_mask = InStorageType(1) << (extra_fraction_len - 1);
-    round = (q & round_mask) != 0;
-    InStorageType sticky_mask = round_mask - 1;
-    sticky = (q & sticky_mask) != 0;
-
-    result = OutFPBits::create_value(
-                 result_sign, 0,
-                 static_cast<OutStorageType>(q >> extra_fraction_len))
-                 .uintval();
-  }
-
-  if (round || sticky)
-    raise_except_if_required(FE_INEXACT);
-
-  bool lsb = (result & 1) != 0;
-
-  switch (quick_get_round()) {
-  case FE_TONEAREST:
-    if (round && (lsb || sticky))
-      ++result;
-    break;
-  case FE_UPWARD:
-    ++result;
-    break;
-  default:
-    break;
-  }
+  if (test_except(FE_OVERFLOW | FE_UNDERFLOW) != 0)
```

Comment:
maybe it is cheaper to test if output is `+Inf` or subnormal/zero?

Reply:
For `FE_UNDERFLOW` we would need a way to check if nonzero bits were truncated off of the mantissa.

Reply:
Zeros are taken cared in your exceptional cases logic earlier, so at this point, the real quotient is non-zero.  Checking the output exponent field should be enough to decide that underflow happens.

Reply:
That's not the only case where we raise `FE_UNDERFLOW`. Remember cases like https://godbolt.org/z/bq6YqGajj. I could go back to only raising `FE_UNDERFLOW` if a nonzero result is rounded to zero if you want, but at this point I'm sure that it's not what IEEE 754 wants us to do.

Reply:
Yes, and in that example, the output `z = 0x1.0p-128` is denormal, it has exponent field == 0:
```
Checking the output exponent field should be enough to decide that underflow happens.
```

Reply:
the only case that we might be called missed with that simple check is when the real value is slightly smaller than `min_normal`, but got rounded to `min_normal`.  And including that case or not is debatable, because the cutting point will depend on the rounding mode.

Reply:
But IEEE 754 states "If the rounded result is exact, no [underflow] flag is raised and no inexact exception is signaled". See https://godbolt.org/z/ohPYnn8Pb. The result is exact, rounding doesn't change it, so no exception is raised.

Reply:
I think the reasoning of IEEE 754 is that underflow might be a problem for an application only because it might get a result that's less precise than usual. So if the result is exact even as a subnormal number then there is no need to warn the application.

Reply:
See my other comment, so that this expensive test can be removed.